src/include/Context.h

```diff
@@ -60,12 +60,10 @@ class Context {
   Context(const Context& other);
   const Context& operator=(const Context& other);
 
- protected:
-  virtual void finish(int r) = 0;
-
  public:
   Context() {}
   virtual ~Context() {}       // we want a virtual destructor!!!
+  virtual void finish(int r) = 0;
```

Comment:
Rather than doing this, it would be better just to just use a function object instead of Context.